README.md
# ConvertTo-Expression
Serializes an object to a PowerShell expression

The ConvertTo-Expression cmdlet converts (serializes) an object to
a PowerShell expression. The object can be stored in a variable,
file or any other common storage for later use or to be ported to
another system.

### Installation

The `ConvertTo-Expression` script can be downloaded from the [PowerShell Gallery](https://www.powershellgallery.com/):
```powershell
Install-Script -Name ConvertTo-Expression
```
As it concerns a standalone script, installation isn't really required. If you don't have administrator rights, you might just download the script (or copy it) to the required location. You might than simply invoke the script using PowerShell [`dot sourcing`](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_scripts?view=powershell-7#script-scope-and-dot-sourcing):
```powershell
. .\ConvertTo-Expression.ps1
```

#### Converting back *from* an expression  
An expression can be restored to an object using the native
`Invoke-Expression` cmdlet:
```powershell
$Object = Invoke-Expression ($Object | ConverTo-Expression)
```
Or Converting it to a `[ScriptBlock]` and invoking it with cmdlets
along with `Invoke-Command` or using the call operator (`&`):
```powershell
$Object = &([ScriptBlock]::Create($Object | ConverTo-Expression))
```
An expression that is stored in a PowerShell (`.ps1`) file might also
be directly invoked by the PowerShell dot-sourcing technique, e.g.:
```powershell
$Object | ConvertTo-Expression | Out-File .\Expression.ps1
$Object = . .\Expression.ps1
```
***Warning:*** Invoking partly trusted input with Invoke-Expression or
`[ScriptBlock]::Create()` methods could be abused by malicious code
injections.

## Examples

Convert a Calendar object to a PowerShell expression:

```powershell
PS C:\> (Get-UICulture).Calendar | ConvertTo-Expression

[pscustomobject]@{
	'AlgorithmType' = 1
	'CalendarType' = 1
	'Eras' = ,1
	'IsReadOnly' = $False
	'MaxSupportedDateTime' = [datetime]'9999-12-31T23:59:59.9999999'
	'MinSupportedDateTime' = [datetime]'0001-01-01T00:00:00.0000000'
	'TwoDigitYearMax' = 2029
}

PS C:\> (Get-UICulture).Calendar | ConvertTo-Expression -Strong

[pscustomobject]@{
	'AlgorithmType' = [System.Globalization.CalendarAlgorithmType]'SolarCalendar'
	'CalendarType' = [System.Globalization.GregorianCalendarTypes]'Localized'
	'Eras' = [array][int]1
	'IsReadOnly' = [bool]$False
	'MaxSupportedDateTime' = [datetime]'9999-12-31T23:59:59.9999999'
	'MinSupportedDateTime' = [datetime]'0001-01-01T00:00:00.0000000'
	'TwoDigitYearMax' = [int]2029
}
```

Save an object in a file and to restore it later:

```powershell
PS C:\>Get-Date | Select-Object -Property * | ConvertTo-Expression | Out-File .\Now.ps1

PS C:\>$Now = .\Now.ps1	# Simular to: $Now = Get-Content .\Now.Ps1 -Raw | Invoke-Expression

PS C:\>$Now

Date        : 1963-10-07 12:00:00 AM
DateTime    : Monday, October 7, 1963 10:47:00 PM
Day         : 7
DayOfWeek   : Monday
DayOfYear   : 280
DisplayHint : DateTime
Hour        : 22
Kind        : Local
Millisecond : 0
Minute      : 22
Month       : 1
Second      : 0
Ticks       : 619388596200000000
TimeOfDay   : 22:47:00
Year        : 1963
```

Compress the PowerShell expression output:

```powershell
PS C:\>@{Account="User01";Domain="Domain01";Admin="True"} | ConvertTo-Expression -Expand -1	

@{'Admin'='True';'Account'='User01';'Domain'='Domain01'}
```

Convert the WinInit Process to a PowerShell expression:

```powershell
PS C:\>WinInitProcess = Get-Process WinInit | ConvertTo-Expression
```
Reveal complex object hierarchies:

```powershell
PS C:\>Get-Host | ConvertTo-Expression -Depth 4
```
## Inputs
`Any`. Each objects provided through the pipeline will converted to an
expression. To concatinate all piped objects in a single expression,
use the unary comma operator, e.g.: `,$Object | ConvertTo-Expression`

## Outputs
`String[]`. ConvertTo-Expression returns a PowerShell expression for
each input object.

## Parameters 

`-InputObject`  
Specifies the objects to convert to a PowerShell expression. Enter
a variable that contains the objects, or type a command or
expression that gets the objects. You can also pipe one or more
objects to `ConvertTo-Expression`.

`-Depth`  
Specifies how many levels of contained objects are included in the 
PowerShell representation. The default value is `9`.

`-Expand`  
Specifies till what level the contained objects are expanded over
separate lines and indented according to the `-Indentation` and 
`-IndentChar` parameters. The default value is `9`.

A negative value will remove redundant spaces and compress the
PowerShell expression to a single line (except for multi-line
strings).

Xml documents and multi-line strings are embedded in a
"here string" and aligned to the left.

`-Indentation`  
Specifies how many IndentChars to write for each level in the hierarchy.

`-IndentChar`  
Specifies which character to use for indenting.

`-Strong`  
By default, the `ConvertTo-Expression` cmdlet will return a weakly typed
expression which is best for transfing objects between differend
PowerShell systems.
The `-Strong` parameter will strickly define value types and objects
in a way that they can still be read by same PowerShell system and
PowerShell system with the same configuration (installed modules etc.).

`-Explore`  
In explore mode, all type prefixes are omitted in the output expression
(objects will cast to to hash tables). In case the `-Strong` parameter is
also supplied, all *orginal* (.Net) type names are shown.
The `-Explore` switch is usefull for exploring object hyrachies and data
type, not for saving and transfering objects.

`-NewLine`  
Specifies which characters to use for a new line. The default is defined by
the operating system.


CHANGELOG
[Version] [DateTime] Author      Comments
--------- ---------- ------      --------
01.00.00  2017-05-31 Ronald Bode First releases
02.10.02  2017-11-21 Ronald Bode Recode along with GetEnumerator.OverloadDefinitions
02.11.10  2017-12-11 Ronald Bode Added support for [DataTable]
02.11.10  2017-12-11 Ronald Bode New Embed function
02.11.10  2017-12-11 Ronald Bode Using Get-Member for object properties
02.12.00  2017-12-12 Ronald Bode Added [Enum] type
02.12.02  2017-12-19 Ronald Bode Add -NewLine parameter
02.12.03  2017-12-19 Ronald Bode Changed -Layers parameter to: -Expand
02.12.04  2017-12-20 Ronald Bode Added support for [Version]
02.12.05  2017-12-20 Ronald Bode Resolved cast bug
02.12.06  2017-12-23 Ronald Bode Changed -SetType to -Type
02.12.07  2017-12-23 Ronald Bode Added -CropText
02.12.08  2018-01-13 Ronald Bode Improved performance by preventing extensive string manipulation
02.12.09  2018-01-15 Ronald Bode Changed -CropText to -ClipText
2.2.0     2018-01-18 Ronald Bode Changed -Type to -TypePrefix
2.2.0     2018-01-18 Ronald Bode Add support for multi-line strings (here string)
2.2.1     2018-01-18 Ronald Bode Removed -ClipText
2.2.2     2018-01-18 Ronald Bode Formatted for PowerShell Gallery and GitHub
2.2.3     2018-01-19 Ronald Bode Removed [Parameter(DontShow)] as it is not downwards compatible with PowerShell 2
2.2.4     2018-01-20 Ronald Bode Added ICONURI
2.2.6     2018-01-20 Ronald Bode Corrected ICONURI
2.2.7     2018-01-27 Ronald Bode Renamed function to: ConvertTo-Expression
2.3.0     2018-01-27 Ronald Bode Renamed solution to: ConvertTo-Expression
2.3.1     2018-01-27 Ronald Bode Changed ConvertTo-PSON to ConvertTo-Expression in Iterate
2.4.0     2018-06-19 Ronald Bode Output an expression [ScriptBlock] rather than a [String]
2.4.1     2018-06-21 Ronald Bode Support for embedded ScriptBlocks
2.4.2     2018-07-15 Ronald Bode Implemented object caching (experimental branch)
2.4.3     2018-07-16 Ronald Bode $Null output in expression [ScriptBlock]
2.4.4     2018-07-18 Ronald Bode Different string manipulation (experimental branch)
2.4.5     2018-08-01 Ronald Bode Return Expression (containing $Null) when input is $Null
2.4.6     2018-08-02 Ronald Bode Preserving embedded empty arrays
2.5.0     2018-08-09 Ronald Bode Using TypeAccelerators and Type.FullName instead of bare Type.Name
2.5.1     2018-08-09 Ronald Bode [Void] unknown types
2.5.2     2018-08-09 Ronald Bode Added [RuntimeTypeHandle]
2.5.3     2018-08-10 Ronald Bode Fixed: ScriptBlock not closed if the last line is a #Remark
2.6.0     2018-08-28 Ronald Bode Fixed empty array `@()` result and simplified recursive calls
2.7.0     2018-08-29 Ronald Bode Directly processing input objects from the pipeline (output is always a single expression)
2.7.0     2018-08-29 Ronald Bode Removed parenthesis `@(...)` from pipeline object list (e.g `'One', 'Two' | ConvertTo-Expression`)
2.7.0     2018-08-29 Ronald Bode Extended numeric type list for (PSv2) typecasting
2.7.0     2018-08-29 Ronald Bode Casting DictionaryEntries and KeyValuePairs to PSObjects (instead of hashtables)
2.7.1     2018-08-29 Ronald Bode Only add a newline to a function when the last line contains a #
2.7.2     2018-08-29 Ronald Bode Default empty type (typecast) for ValueTypes
2.7.3     2018-11-16 Ronald Bode Changed [PSObject] to [PSCustomObject] type
2.7.4     2018-11-28 Ronald Bode Fixed IntPtr output type
2.7.5     2018-11-29 Ronald Bode Preventing nested array flattening by adding a comma
3.0.0     2018-11-29 Ronald Bode Minimized root top (array) expression top
3.0.1     2018-11-30 Ronald Bode Minimized all array expressions (only use @ sign and brackets were required)
3.0.2     2018-12-14 Ronald Bode Fixed PSScript Analyzer PSUseDeclaredVarsMoreThanAssignment issue
3.0.3     2018-12-14 Ronald Bode Renamed option: -TypePrefix Strict to: -TypePrefix Strong
3.0.4     2018-12-14 Ronald Bode Cut off recurring references with: $Self (followed by the concerned number of .Parents properties)
3.0.5     2019-01-05 Ronald Bode Cut off recurring references with path to the referenced property: $_[.<property>]
3.0.9     2019-01-05 Ronald Bode Removed PSON references and aliases
3.1.0     2019-01-05 Ronald Bode Replaced -TypePrefix parameter with -Strong and -Explore switch
3.1.1     2019-01-07 Ronald Bode Fixed embedded PSCustomObject issue due to recurring reference tracking
3.2.0     2019-05-14 Ronald Bode Recode better recognizing object containers
3.2.1     2019-05-21 Ronald Bode Using unary comma operator
3.2.2     2019-05-21 Ronald Bode List arrays without brackets where possible
3.2.3     2019-05-28 Ronald Bode Removed extra linefeed after here string
3.2.5     2019-06-04 Ronald Bode Convert arrays of DictionaryEntries and arrays of KeyValuePairs to HashTable
3.2.7     2019-06-11 Ronald Bode Used PowerShell Accelerators rather than .Net
3.2.8     2019-06-18 Ronald Bode Handling streamed object as individual objects
3.2.9     2019-06-25 Ronald Bode No trailing space in any lines
3.2.10    2019-06-25 Ronald Bode Preserve object property orders
3.2.11    2019-08-27 zcla        Resolved bug #1: Single quote in Hashtable key
3.2.12    2019-08-29 Ronald Bode Support for SemVer type
3.2.13    2019-08-29 Ronald Bode Using .isPrimitive property rather than numeric type table 
3.2.15    2019-10-18 Ronald Bode Support for valued arrays (use -Strong)
3.2.17    2019-10-19 Ronald Bode Added support for 'mailaddress', 'Regex', 'Type', 'Version', 'Uri'
3.2.19    2019-10-19 Ronald Bode Support Credential and SecureString
3.2.20    2020-03-20 Ronald Bode Removed (unused) -Concatenate switch
3.3.0     2020-03-20 Ronald Bode Removed all statements that could exploit for code injection
3.3.1     2020-08-26 Ronald Bode Fixed NullReferenceException
3.3.2     2020-08-27 Ronald Bode Using `.psobject.methods['<method name>'] -is [PSMethod]` (slightly faster)

Tests/ConvertFrom-Table.ps1
Function ConvertFrom-Table {
	Param (
		[Parameter(ValueFromPipeLine = $True)][String[]]$Table, 
		[Char]$HeaderChar = '-', [Char[]]$ArrayMask = '{,}', [Char[]]$ExpressionMask = '`'
	)
	Begin {
		Function Get-Values ($Row) {
			For ($i = 0; $i -lt $Index.Count; $i++) {
				$Start = $Index[$i]
				$End = If ($i -lt $Index.Count - 1) {$Index[$i + 1]} Else {$Row.Length}
				If($End -gt $Row.Length) {$End = $Row.Length}
				If ($Start -lt $End) {$Row.SubString($Start, $End - $Start).TrimEnd()} Else {""}
			}
		}
		$HeaderRow = $Null
		$Property = @{}
		$Name = $Null
		$Index = @()
	}
	Process {
		$Table | ForEach-Object {
			ForEach ($Row in ($_ -Split "[\r\n]+")) {
				If ($Name) {
					$Value = Get-Values $Row
					For ($i = 0; $i -lt $Name.Count; $i++) {
						$Property.($Name[$i]) = 
						If ($ArrayMask[1] -and $Value[$i].StartsWith($ArrayMask[0]) -And $Value[$i].EndsWith($ArrayMask[2])) {
							$Value[$i].SubString(1, $Value[$i].Length - $ArrayMask[2].length - 1) -Split $ArrayMask[1] | ForEach-Object {$_.Trim()}
						} ElseIf ($ExpressionMask -and $Value[$i].StartsWith($ExpressionMask[0]) -And $Value[$i].EndsWith($ExpressionMask[1])) {
							Invoke-Expression $Value[$i].SubString(1, $Value[$i].Length - $ExpressionMask[1].length - 1)
						} Else {$Value[$i]}
					}
					New-Object PSObject -Property $Property
				} Else {
					If ($HeaderRow) {
						$Index = Select-String "(?<!\S)[\x2D]+" -Input $Row -AllMatches | ForEach-Object {$_.Matches} | Select-Object -Expand Index
						If ($Index.Count) {$Name = Get-Values $HeaderRow}
					}
					$HeaderRow = $Row
				}
			}
		}
	}
	End {
		If (!$Name) {Write-Error "No table header found"}
	}
}


ConvertTo-Pson.ps1
<#PSScriptInfo
.VERSION 2.2.6
.GUID 5f167621-6abe-4153-a26c-f643e1716720
.AUTHOR Ronald Bode (iRon)
.DESCRIPTION Serializes an object to a PowerShell expression.
.COMPANYNAME 
.COPYRIGHT 
.TAGS PSON PowerShell Object Notation expression serialize
.LICENSEURI https://github.com/iRon7/ConvertTo-PSON/LICENSE.txt
.PROJECTURI https://github.com/iRon7/ConvertTo-PSON
.ICONURI https://raw.githubusercontent.com/iRon7/ConvertTo-PSON/master/PsonIcon.png
.EXTERNALMODULEDEPENDENCIES 
.REQUIREDSCRIPTS 
.EXTERNALSCRIPTDEPENDENCIES 
.RELEASENOTES
.PRIVATEDATA 
#>

Function ConvertTo-Pson {
	<#
		.SYNOPSIS
			Serializes an object to a PowerShell expression.

		.DESCRIPTION
			The ConvertTo-Pson cmdlet converts any object to a string in PowerShell Object
			Notation (PSON) format. The properties are converted to field names, the field
			values are converted to property values, and the methods are removed.
			You can then use the ConvertFrom-Pson (Invoke-Expression) cmdlet to convert a
			PSON-formatted string to a PowerShell object, which is easily managed in
			Windows PowerShell.

		.PARAMETER InputObject
			Specifies the objects to convert to a PSON expression. Enter a variable that
			contains the objects, or type a command or expression that gets the objects.
			You can also pipe an object to ConvertTo-Pson.

		.PARAMETER Depth
			Specifies how many levels of contained objects are included in the PSON
			representation. The default value is 9.

		.PARAMETER Expand
			Specifies till what level the contained objects are expanded over separate lines
			and indented according to the -Indentation and -IndentChar parameters.
			The default value is 9.
			
			A negative value will remove redundant spaces and compress the PSON expression to
			a single line (except for multiline strings).
			
			Xml documents and multiline strings are embedded in a "here string" and aligned
			to the left.
			
		.PARAMETER Indentation
			Specifies how many IndentChars to write for each level in the hierarchy.

		.PARAMETER IndentChar
			Specifies which character to use for indenting.

		.PARAMETER Type
			Defines how the explicite the object type is being parsed:

			-Type None
				No type information will be added to the (embedded) objects and values in
				the PSON string. This means that objects and values will be parsed to one
				of the following data types when reading them back with ConvertFrom-Pson
				(Invoke-Expression): a numeric value, a [String] ('...'), an [Array] 
				(@(...)) or a [HashTable] (@{...}).

			-Type Native
				The original type prefix is added to the (embedded) objects and values in
				the PSON string. Note that most system (.Net) objects can’t be read back
				with ConvertFrom-Pson (Invoke-Expression), but -SetType Name can help to
				reveal (embedded) object types and hierarchies.

			-Type Cast (Default)
				The type prefix is only added to (embedded) objects and values when required
				and optimized for read back with ConvertFrom-Pson (Invoke-Expression) by e.g.
				converting system (.Net) objects to PSCustomObject objects. Numeric values
				won't have a strict type and therefor parsed to the default type that fits
				the value when read back with ConvertFrom-Pson (Invoke-Expression).

			-Type Strict
				All (embedded) objects and values will have an explicit type prefix optimized
				for read back with ConvertFrom-Pson (Invoke-Expression) by e.g. converting
				system (.Net) objects to PSCustomObject objects.

		.PARAMETER NewLine
			Specifies which characters to use for a new line. The default is defined by the
			operating system.

		.PARAMETER Iteration
			Do not use (for internal use only).

			.EXAMPLE 

			PS C:\>(Get-UICulture).Calendar | ConvertTo-Pson	# Convert a Calendar object to a PowerShell expression

			[PSCustomObject]@{
				'AlgorithmType' = 'SolarCalendar'
				'CalendarType' = 'Localized'
				'Eras' = 1
				'IsReadOnly' = $False
				'MaxSupportedDateTime' = [DateTime]'9999-12-31T23:59:59.9999999'
				'MinSupportedDateTime' = [DateTime]'0001-01-01T00:00:00.0000000'
				'TwoDigitYearMax' = 2029
			}

		.EXAMPLE 

			PS C:\>@{Account="User01";Domain="Domain01";Admin="True"} | ConvertTo-Pson -Expand -1	# Compress the PSON output

			@{'Admin'='True';'Account'='User01';'Domain'='Domain01'}


		.EXAMPLE 

			PS C:\>Get-Date | Select-Object -Property * | ConvertTo-Pson	# Convert an object to a PSON expression and to a PowerShell object

			[PSCustomObject]@{
				'Date' = [DateTime]'2018-01-09T00:00:00.0000000+01:00'
				'DateTime' = 'Tuesday, January 9, 2018 7:22:57 PM'
				'Day' = 9
				'DayOfWeek' = 'Tuesday'
				'DayOfYear' = 9
				'DisplayHint' = 'DateTime'
				'Hour' = 19
				'Kind' = 'Local'
				'Millisecond' = 671
				'Minute' = 22
				'Month' = 1
				'Second' = 57
				'Ticks' = 636511225776716485
				'TimeOfDay' = [TimeSpan]'19:22:57.6716485'
				'Year' = 2018
			}

			PS C:\>Get-Date | Select-Object -Property * | ConvertTo-Pson | ConvertFrom-Pson

			Date        : 2018-01-09 12:00:00 AM
			DateTime    : Tuesday, January 9, 2018 7:27:43 PM
			Day         : 9
			DayOfWeek   : Tuesday
			DayOfYear   : 9
			DisplayHint : DateTime
			Hour        : 19
			Kind        : Local
			Millisecond : 76
			Minute      : 27
			Month       : 1
			Second      : 43
			Ticks       : 636511228630764893
			TimeOfDay   : 19:27:43.0764893
			Year        : 2018

		.EXAMPLE 

			PS C:\>WinInitProcess = Get-Process WinInit | ConvertTo-Pson	# Convert the WinInit Process to a PSON expression

		.EXAMPLE 

			PS C:\>Get-Host | ConvertTo-Pson -Depth 4	# Reveal complex object hierarchies

		.LINK
			Invoke-Expression (Alias ConvertFrom-Pson)
	#>
	[CmdletBinding()][OutputType([String])]Param (
		[Parameter(ValueFromPipeLine = $True)][Object[]]$InputObject, [Int]$Depth = 9, [Int]$Expand = 9,
		[Int]$Indentation = 1, [String]$IndentChar = "`t", [ValidateSet("None", "Native", "Cast", "Strict")][String]$TypePrefix = "Cast",
		[String]$NewLine = [System.Environment]::NewLine, [Int]$Iteration = 0
	)
	$PipeLine = $Input | ForEach-Object {$_}; If ($PipeLine) {$InputObject = $PipeLine}
	Function Iterate ($Value) {ConvertTo-Pson @(,$Value) $Depth $Expand $Indentation $IndentChar $TypePrefix $NewLine ($Iteration + 1)}
	Function Embed ($List, $Dictionary) {If ($Iteration -ge $Depth) {If ($Null -ne $Dictionary) {Return "@{}"} Else {Return "@()"}}
		$Items = ForEach ($Key in $List) {If ($Null -ne $Dictionary) {"'$Key'$Space=$Space" + (Iterate $Dictionary.$Key)} Else {Iterate $Key}}
		$Open, $Join, $Separator, $Close = If ($Null -ne $Dictionary) {"@{", ";$Space", "$LineUp$Tab", "}"} Else {"@(", ",$Space", ",$LineUp$Tab", ")"}
		$Open + (&{If (($Iteration -ge $Expand) -or (@($Items).Count -le 1)) {$Items -Join $Join} Else {"$LineUp$Tab$($Items -Join $Separator)$LineUp"}}) + $Close
	}
	$Object = If (@($InputObject).Count -eq 1) {@($InputObject)[0]} Else {$InputObject}
	If ($Null -eq $Object) {"`$Null"} Else {
		$Space = If ($Iteration -gt $Expand) {""} Else {" "}; $Tab = $IndentChar * $Indentation; $LineUp = "$NewLine$($Tab * $Iteration)"
		$Type = $Object.GetType().Name; $Cast = $Null; $Enumerator = $Object.GetEnumerator.OverloadDefinitions
		$PSON = If ($Object -is [Boolean]) {If ($Object) {'$True'} Else {'$False'}}
		ElseIf ($Object -is [Char]) {$Cast = $Type; "'$Object'"}
		ElseIf ($Object -is [String]) {If ($Object -Match "[`r`n]") {"@'$NewLine$Object$NewLine'@$NewLine"} Else {"'$($Object.Replace('''', ''''''))'"}}
		ElseIf ($Object -is [DateTime]) {$Cast = $Type; "'$($Object.ToString('o'))'"}
		ElseIf ($Object -is [TimeSpan] -or $Object -is [Version]) {$Cast = $Type; "'$Object'"}
		ElseIf ($Object -is [Enum]) {$Type = "String"; "'$($Object)'"}
		ElseIf ($Object -is [Xml]) {$Cast = "Xml"; $SW = New-Object System.IO.StringWriter; $XW = New-Object System.Xml.XmlTextWriter $SW
			$XW.Formatting = If ($Level -gt $Expand) {"None"} Else {"Indented"}; $XW.Indentation = $Indentation; $XW.IndentChar = $IndentChar
			$Object.WriteContentTo($XW); If ($Level -gt $Expand) {"'$SW'"} Else {"@'$NewLine$SW$NewLine'@$NewLine"}}
		ElseIf ($Object.GetType().Name -eq "DictionaryEntry" -or $Type -like "KeyValuePair*") {$Type = "Hashtable"; Embed $Object.Key @{$Object.Key = $Object.Value}}
		ElseIf ($Object.GetType().Name -eq "OrderedDictionary") {$Type = "Hashtable"; $Cast = "Ordered"; Embed $Object.Keys $Object}
		ElseIf ($Enumerator -match "[\W]IDictionaryEnumerator[\W]") {$Type = "Hashtable"; Embed $Object.Keys $Object}
		ElseIf ($Enumerator -match "[\W]IEnumerator[\W]" -or $Object.GetType().Name -eq "DataTable") {$Type = "Array"; Embed $Object}
		Else {$Property = $Object | Get-Member -Type Property; If (!$Property) {$Property = $Object | Get-Member -Type NoteProperty}
			$Names = ForEach ($Name in ($Property | Select-Object -Expand "Name")) {$Object.PSObject.Properties |
				Where-Object {$_.Name -eq $Name -and $_.IsGettable} | Select-Object -Expand "Name"}
			If ($Property) {$Type = "PSCustomObject"; $Cast = $Type; Embed $Names $Object} Else {$Object}
		}
		Switch ($TypePrefix) {
			'None'  	{"$PSON"}
			'Native'	{"[$($Object.GetType().Name)]$PSON"}
			'Cast'  	{If ($Cast) {"[$Cast]$PSON"} Else {"$PSON"}}
			'Strict'	{If ($Cast) {"[$Cast]$PSON"} Else {"[$Type]$PSON"}}
		}
	}
} Set-Alias PSON ConvertTo-Pson -Description "Convert variable to PSON"
Set-Alias ConvertFrom-Pson Invoke-Expression -Description "Convert object from PSON"
